File: python/translator.py
import sys

# Braille dictionary
BRAILLE_LETTERS = {
    'a': 'O.....', 'b': 'O.O...', 'c': 'OO....', 'd': 'OO.O..', 'e': 'O..O..',
    'f': 'OOO...', 'g': 'OOOO..', 'h': 'O.OO..', 'i': '.OO...', 'j': '.OOO..',
    'k': 'O...O.', 'l': 'O.O.O.', 'm': 'OO..O.', 'n': 'OO.OO.', 'o': 'O..OO.',
    'p': 'OOO.O.', 'q': 'OOOOO.', 'r': 'O.OOO.', 's': '.OO.O.', 't': '.OOOO.',
    'u': 'O...OO', 'v': 'O.O.OO', 'w': '.OOO.O', 'x': 'OO..OO', 'y': 'OO.OOO',
    'z': 'O..OOO', ' ': '......','capital': '.....O', 'number': '.O.OOO'
}
BRAILLE_NUMBERS = {
    '1': 'O.....', '2': 'O.O...', '3': 'OO....', '4': 'OO.O..', '5': 'O..O..',
    '6': 'OOO...', '7': 'OOOO..', '8': 'O.OO..', '9': '.OO...', '0': '.OOO..'
    }
# Reverse Braille to English
ENGLISH_LETTERS = {v: k for k, v in BRAILLE_LETTERS.items()}

ENGLISH_NUMBERS = {v: k for k, v in BRAILLE_NUMBERS.items()}

# Check if input is Braille or English
def is_braille(input_str):
    if not isinstance(input_str, str):
        raise ValueError("Input must be a string")
    return len(input_str) % 6 == 0 and set(input_str).issubset({'O', '.'})


def braille_to_english(braille_input):
    result = []
    index = 0
    number_mode = False
    while index < len(braille_input):
        braille_char = braille_input[index:index+6]
        if braille_char == BRAILLE_LETTERS['capital']:
            index += 6
            braille_char = braille_input[index:index+6]
            result.append(ENGLISH_LETTERS[braille_char].upper())
        elif braille_char == BRAILLE_LETTERS['number']:
            number_mode = True
        elif number_mode and braille_char in ENGLISH_NUMBERS:
            result.append(ENGLISH_NUMBERS[braille_char])
        else:
            result.append(ENGLISH_LETTERS.get(braille_char, '?'))
        index += 6
    return ''.join(result)

# Translate English to Braille
def english_to_braille(english_input):
    result = []
    number_mode = False

    for char in english_input:
        if char.isdigit():
            if not number_mode:
                result.append(BRAILLE_LETTERS['number'])
                number_mode = True
            result.append(BRAILLE_NUMBERS[char])
        else:
            if char.isupper():
                result.append(BRAILLE_LETTERS['capital'])
                result.append(BRAILLE_LETTERS[char.lower()])
            else:
                result.append(BRAILLE_LETTERS[char])
            number_mode = False

    return ''.join(result)

def main():
    # Get the input arguments (concatenate them if multiple words are passed)
    input_str = ' '.join(sys.argv[1:])
    
    # Check input and do appropriate translation
    if is_braille(input_str):
        output = braille_to_english(input_str)
    else:
        output = english_to_braille(input_str)
    # Print the output
    print(output)

if __name__ == "__main__":
    main()

File: javascript/translator.js
// Braille to English mapping
const brailleToEnglishAlph = {
	"O.....": "a",
	"O.O...": "b",
	"OO....": "c",
	"OO.O..": "d",
	"O..O..": "e",
	"OOO...": "f",
	"OOOO..": "g",
	"O.OO..": "h",
	".OO...": "i",
	".OOO..": "j",
	"O...O.": "k",
	"O.O.O.": "l",
	"OO..O.": "m",
	"OO.OO.": "n",
	"O..OO.": "o",
	"OOO.O.": "p",
	"OOOOO.": "q",
	"O.OOO.": "r",
	".OO.O.": "s",
	".OOOO.": "t",
	"O...OO": "u",
	"O.O.OO": "v",
	".OOO.O": "w",
	"OO..OO": "x",
	"OO.OOO": "y",
	"O..OOO": "z",
};

const capital = ".....O";
const number = ".O.OOO";
const space = "......";

const brailleToEnglishNum = {
	".OOO..": "0",
	"O.....": "1",
	"O.O...": "2",
	"OO....": "3",
	"OO.O..": "4",
	"O..O..": "5",
	"OOO...": "6",
	"OOOO..": "7",
	"O.OO..": "8",
	".OO...": "9",
};

// English to Braille mapping
// Initialize an empty object to store English to Braille mappings
const englishToBraille = {};

// Loop over each entry in the brailleToEnglish letters object
Object.entries(brailleToEnglishAlph).forEach(([braille, char]) => {

// Handle for uppercase letters
    englishToBraille[char] = braille;
    englishToBraille[char.toUpperCase()] = capital + braille;
});

// Handle numbers
Object.entries(brailleToEnglishNum).forEach(([braille, char]) => {
		englishToBraille[char] = braille;
	
});

// Space character mapping
englishToBraille[" "] = space;


// Check if input is Braille (if input contains only 'O' or '.' characters)
function isBraille(input) {
	return /^[O.]+$/.test(input);
}
// Check if character is number 
function isNumber(char) {
    return /\d/.test(char);
}

// Translate English to Braille
function translateEnglishToBraille(input) {
	let output = "";
    let numMode = false;

	for (let i = 0; i < input.length; i++) {   //handle numbers
		const char = input[i];
        if (isNumber(char) && !numMode) {
            output += number + englishToBraille[char]
            numMode = true
        } else if  (isNumber(char) && numMode) {
            output += englishToBraille[char]
        } else {
            numMode = false
            output += englishToBraille[char]
        }
	}
	return output;
}

// Translate Braille to English
function translateBrailleToEnglish(input) {
	let output = "";
	let i = 0;
	let isCapital = false;
	let isNumber = false;

	while (i < input.length) {
		const currentSymbol = input.slice(i, i + 6); // to account for 6 braille symbols
		if (currentSymbol === capital) {        //to account for uppercase
			isCapital = true;
		} else if (currentSymbol === number) { // to account for numbers
			isNumber = true;
		} else if (currentSymbol === space) { // to account for space
			output += " "; 
			isNumber = false;
		} else if (brailleToEnglishAlph[currentSymbol]) { 
			let char = brailleToEnglishAlph[currentSymbol]; 
			if (isCapital) {
				output += char.toUpperCase();
				isCapital = false;
			} else if (isNumber) {
				let num = brailleToEnglishNum[currentSymbol];
				output += num;
			} else if (!isNumber) {
				output += char;
			}
		}
		i += 6;
	}
	return output;
}

// Main traslator function
function mainTranslator() {
	const commandInput = process.argv.slice(2); //  slicing the first two arguments to command line argument
	const input = commandInput.join(" ");
	let output = "";

	if (isBraille(input)) {
		if (input.length % 6 === 0) {
			output = translateBrailleToEnglish(input);
		} else {
			output = "Error: incorrect Braille patterns";
		}
	} else {
		output = translateEnglishToBraille(input);
	}
	console.log(output);
}

mainTranslator();


